src/inference/src/dev/threading/cpu_streams_executor.cpp

```diff
@@ -341,6 +339,17 @@ struct CPUStreamsExecutor::Impl {
             : ThreadLocal<std::shared_ptr<Stream>>(callback_construct),
               _impl(impl) {}
         std::shared_ptr<Stream> local() {
+            // maybe there are two CPUStreamsExecutors in the same thread.
+            static thread_local std::map<void*, std::shared_ptr<std::thread::id>> t_stream_count_map;
+            // fix the memory leak issue that CPUStreamsExecutor is already released,
+            // but still exists thread id in t_stream_count_map
+            for (auto it = t_stream_count_map.begin(); it != t_stream_count_map.end();) {
+                if (this != it->first && it->second.use_count() == 1) {
+                    t_stream_count_map.erase(it++);
+                } else {
+                    it++;
+                }
+            }
```

Comment:
Actually this is not a really reliable solution since it introduces coupling between this class and some external behavior (threads management). Looks like a kind of a standard problem of avoiding extra ownership of a resource. Why can't we use a weak_ptr here `std::map<void*, std::weak_ptr<std::thread::id>>`? Another question is why do we use bin tree when the key type is a 8 bit integer? Isn't it better to use hash table instead?

Reply:
Yes I agree with you，the root issue is that cpuStreamExecuter need  Impl::Stream instance,  but Impl::Stream instance has two owner, one is thread, and one is cpuStreamExecuter, one of the ower is destroyed, impl::stream should be destroyed.
the origin code use tbb thread local class to store Impl::Stream, tbb thread local is a member of cpuStreamExecuter, but when customer thread is destroyed, tbb thread local don't know it, the memory leak issue happen, this is origin pr https://github.com/openvinotoolkit/openvino/pull/19191.  At the start of the PR 19191, I try to use c++ key word thread local to store Impl::Stream, but there are lifecycle issue, cpuStreamExecuter is destroy but Impl::Stream is still in thread local.
return back to the question.
1 I can't use std::weak_ptr here, it will lost the connection between impl::stream and thread.
2 about hash table and map, below is just my opinion, maybe wrong 
the key of map is the pointer of cpuStreamExecuter::Impl, I think there will not be many cpuStreamExecuter in same thread, so I am not sure hash table is more efficient here, hash table maybe use more memory than map and customer may has many threads. 
sorry I didn't get the point of relation between 8 byte and hash table

Reply:
> 1 I can't use std::weak_ptr here, it will lost the connection between impl::stream and thread.

How exactly it loses connection? It looks like the only reason of having a pointer to `std::thread::id` is tracking dynamically changing value, but with a guaranty that it's valid. So why do we need the ownership?

> 2 about hash table and map, below is just my opinion, maybe wrong
the key of map is the pointer of cpuStreamExecuter::Impl, I think there will not be many cpuStreamExecuter in same thread, so I am not sure hash table is more efficient here, hash table maybe use more memory than map and customer may has many threads.

Ok, when the amount of elements is low, we can use bin tree.

>sorry I didn't get the point of relation between 8 byte and hash table

Hash function already exists and trivial.

Reply:
before PR19191, thread didn't hold impl::stream, so it will cause memory leak when create and destroy thread frequently
```
thread -------------------------|
                                |
                             not hold
                                |
                                v
cpuStreamExecuter ---hold---> impl::stream
```
after PR 19191, if the thread is destroyed, the count of `std::shared_ptr<std::thread::id>` will became 1 and  when  call local(), cpuStreamExecuter  will check the count of `std::shared_ptr<std::thread::id>`, if it became 1, cpuStreamExecuter will    reuse  `impl::stream` or destroy `impl::stream`
```
thread ----shared_ptr--hold -> std::thread::id
                                       ^
                                       |
    |------shared_ptr-----hold-------- |                 
    |					
cpuStreamExecuter ---hold---> impl::stream
```

Reply:
Relaying on the `std::shared_ptr ` ref counter is a really bad idea. Suppose at some point of time, in some place in the code yet another copy of the shared_ptr happens, then this assumption is broken and we will run into hardly detecting memory leak again. We need a better solution.

Reply:
@maxnick  if you have new idea or solution, I think we can talk with wanglei.
about how `std::shared_ptr<std::thread::id>` controls `impl::stream` lifetime, 
cpuStreamExecuter  use  member  `_stream_map` to hold `impl::stream`,  and the key of ` _stream_map` is the shared_ptr of `std::thread::Id`, the value of `_stream_map` is the ` impl::stream`. if the count of `std::shared_ptr<std::thread::id>`  became 1, 
local() will reuse `impl::stream` or destroy `impl::stream`, please check the code https://github.com/openvinotoolkit/openvino/pull/19191
```
thread ----shared_ptr--hold -> std::thread::id
                                       ^
                                       |	
                                    shared_ptr
                                       |		
cpuStreamExecuter --------_stream_map<key ,value>---> impl::stream
```

Reply:
Ok, I see. But I'm still against using the std::shared_ptr ref counter as a condition to perform any actions. As I described above this counter is out of the implementation control so an error may be easily introduced.
If I correctly understand your explanation, the problem may be formulated as follows:
We need to have a mechanism which releases `impl::stream` object when the corresponding thread is destroyed. To this end you may employ some RAII technics, or indeed some sort of an atomic counter, but it shall be protected from unintentional modifications (in in opposite to `std::shared_ptr` ref counter which can be modified by coping the object).

Reply:
> We need to have a mechanism which releases impl::stream object when the corresponding thread is destroyed

yes，it is the PR [19191](https://github.com/openvinotoolkit/openvino/pull/19191) want to fix 

> std::shared_ptr ref counter which can be modified by coping the object

`std::shared_ptr<std::thread::id>` only used in local() function and t_stream_count_map.  t_stream_count_map is also defined in local(), so the count of `std::shared_ptr<std::stread::id>` only can be modified by local() function. if local() do well, there will be no issue. if there is still concern on it, I think I can add assert for it, how do you think ?
```suggestion
            for (auto it = t_stream_count_map.begin(); it != t_stream_count_map.end();) {
                if (this != it->first && it->second.use_count() == 1) {
                    t_stream_count_map.erase(it++);
                } else {
                    openvino_assert(it->second.use_count() <= 2, 
                            "the count should not be greater than 2")
                    it++;
                }
            }
```

> RALL technics

could you give more detail about this technics, what is RALL technics?

Reply:
> could you give more detail about this technics, what is RALL technics?

https://en.cppreference.com/w/cpp/language/raii

Reply:
@maxnick  @wangleis I implement  `CustomThreadLocal ::ThreadID` which is a simple version of  `shard_ptr<std::thread::id>`, please help to continue to review